substrate/frame/beefy/src/equivocation.rs

```diff
@@ -239,49 +311,37 @@ where
 /// unsigned equivocation reports.
 impl<T: Config> Pallet<T> {
 	pub fn validate_unsigned(source: TransactionSource, call: &Call<T>) -> TransactionValidity {
-		if let Call::report_equivocation_unsigned { equivocation_proof, key_owner_proof } = call {
-			// discard equivocation report not coming from the local node
-			match source {
-				TransactionSource::Local | TransactionSource::InBlock => { /* allowed */ },
-				_ => {
-					log::warn!(
-						target: LOG_TARGET,
-						"rejecting unsigned report equivocation transaction because it is not local/in-block."
-					);
-					return InvalidTransaction::Call.into()
-				},
-			}
-
-			let evidence = (*equivocation_proof.clone(), key_owner_proof.clone());
-			T::EquivocationReportSystem::check_evidence(evidence)?;
-
-			let longevity =
-				<T::EquivocationReportSystem as OffenceReportSystem<_, _>>::Longevity::get();
-
-			ValidTransaction::with_tag_prefix("BeefyEquivocation")
-				// We assign the maximum priority for any equivocation report.
-				.priority(TransactionPriority::MAX)
-				// Only one equivocation report for the same offender at the same slot.
-				.and_provides((
-					equivocation_proof.offender_id().clone(),
-					equivocation_proof.set_id(),
-					*equivocation_proof.round_number(),
-				))
-				.longevity(longevity)
-				// We don't propagate this. This can never be included on a remote node.
-				.propagate(false)
-				.build()
-		} else {
-			InvalidTransaction::Call.into()
+		// discard equivocation report not coming from the local node
+		match source {
+			TransactionSource::Local | TransactionSource::InBlock => { /* allowed */ },
+			_ => {
+				log::warn!(
+					target: LOG_TARGET,
+					"rejecting unsigned report equivocation transaction because it is not local/in-block."
+				);
+				return InvalidTransaction::Call.into()
+			},
 		}
+
+		let evidence = call.to_equivocation_evidence_for().ok_or(InvalidTransaction::Call)?;
+		let tag = (evidence.offender_id().clone(), evidence.set_id(), *evidence.round_number());
```

Comment:
Do we need to accept two different types of equivocation reports (double voting and fork voting) at a single slot? If so, we maybe need to add that type to the tag

Reply:
I don't think we can accept 2 equivocation reports from the same offender in the same round. I think `check_evidence` will fail:

```
	fn check_evidence(
		evidence: EquivocationEvidenceFor<T>,
	) -> Result<(), TransactionValidityError> {
		let offender = evidence.checked_offender::<P>().ok_or(InvalidTransaction::BadProof)?;

		// Check if the offence has already been reported, and if so then we can discard the report.
		let time_slot = TimeSlot { set_id: evidence.set_id(), round: *evidence.round_number() };
		if R::is_known_offence(&[offender], &time_slot) {
			Err(InvalidTransaction::Stale.into())
		} else {
			Ok(())
		}
	}
```

Reply:
Or we could change `check_evidence`. I don't know what would be best. I would lean towards staying consistent with the current approach where we accept one equivocation per `(offender, set_id, round)`.

Reply:
I don't see anywhere in this PR the extrinsic used to report these ForkEquivocations..

I am thinking since you only allow single offender per report does that mean that for a slashable fork equivocation we need to report all the offenders across multiple blocks?

Reply:
> I don't see anywhere in this PR the extrinsic used to report these ForkEquivocations..

The extrinsic is [here](https://github.com/serban300/polkadot-sdk/blob/4b69c20fdd5de77b043c5fc601b4a9d111a94645/substrate/frame/beefy/src/lib.rs#L297) .

> I am thinking since you only allow single offender per report does that mean that for a slashable fork equivocation we need to report all the offenders across multiple blocks?

Hmmm good point. Could be. I can try to implement the solution that was discussed [here](https://github.com/paritytech/polkadot-sdk/pull/1903#discussion_r1538896995). It's good to have it anyway.

Reply:
> The extrinsic is [here](https://github.com/serban300/polkadot-sdk/blob/4b69c20fdd5de77b043c5fc601b4a9d111a94645/substrate/frame/beefy/src/lib.rs#L297) .

Right, I got confused. I had noticed we don't have runtime APIs for the "unsigned" version of the call. Should we add it?

Reply:
Yes, missed that. Will add it.

Reply:
> > I am thinking since you only allow single offender per report does that mean that for a slashable fork equivocation we need to report all the offenders across multiple blocks?
> 
> Hmmm good point. Could be. I can try to implement the solution that was discussed [here](https://github.com/paritytech/polkadot-sdk/pull/1903#discussion_r1538896995). It's good to have it anyway.

Done. PTAL !



>> Right, I got confused. I had noticed we don't have runtime APIs for the "unsigned" version of the call. Should we add it?

> Yes, missed that. Will add it.

On a second thought, I think it's better to do this in a separate PR together with the logic for generating the ancestry proofs since the type of the ancestry proof will depend on that (it could either be `AncestryProof` or `OpaqueValue`, depending on how we generate it). WDYT ?